packages/x-charts/src/ScatterChart/formatter.ts

```diff
@@ -1,9 +1,55 @@
-import { defaultizeValueFormatter } from '../internals/defaultizeValueFormatter';
 import { SeriesFormatter } from '../context/PluginProvider/SeriesFormatter.types';
+import { ScatterValueType } from '../models';
+
+const formatter: SeriesFormatter<'scatter'> = ({ series, seriesOrder }, dataset) => {
+  const completeSeries = Object.fromEntries(
+    Object.entries(series).map(([seriesId, seriesData]) => {
+      const xDataKey = seriesData?.datasetKeys?.x;
+      const yDataKey = seriesData?.datasetKeys?.y;
+      const zDataKey = seriesData?.datasetKeys?.z;
+      const idDataKey = seriesData?.datasetKeys?.id;
+
+      const keys = [xDataKey, yDataKey, idDataKey];
+      const hasStringKeys = keys.some((key) => typeof key === 'string');
+      const hasNonStringKeys = keys.some((key) => typeof key !== 'string');
+      if (hasStringKeys && hasNonStringKeys) {
+        throw new Error(
+          [
+            `MUI X: scatter series with id='${seriesId}' has inconsistent data keys.`,
+            'Either provide all data keys or none.',
+          ].join('\n'),
+        );
+      }
+
+      const data = !hasStringKeys
+        ? (seriesData.data ?? [])
+        : (dataset?.map((d) => {
+            const x = d[xDataKey!]!;
+            const y = d[yDataKey!]!;
+            const z = d[zDataKey ?? ''];
+            const id = d[idDataKey!]!;
+
```

Comment:
Since `datasetKeys` should contain x, y, and id if defined, some `!` could be removed
```suggestion
      const data = !seriesData?.datasetKeys
        ? (seriesData.data ?? [])
        : (dataset?.map((d) => {
            const x = d[seriesData?.datasetKeys.x];
            const y = d[seriesData?.datasetKeys.y];
            const z = seriesData?.datasetKeys && d[seriesData?.datasetKeys.z];
            const id = d[seriesData?.datasetKeys.id];

```

Reply:
Its kind of required still 

```tsx
            const x = d[seriesData?.datasetKeys!.x];
            const y = d[seriesData?.datasetKeys!.y];
            const z = seriesData?.datasetKeys?.z && d[seriesData?.datasetKeys!.z!];
            const id = d[seriesData?.datasetKeys!.id];
```

Reply:
I've made it closer to my initial solution, but made it cleaner, so we don't need all those `!` 😅

Reply:
I added a commit to extract the `datasetKeys` to avoid all those `seriesData?.datasetKeys`. And instead of getting a boolean, I get the list of missing properties to have a more precise error message.

I also updated the docs demo to avoid overflow between y-axis ticks and the label